File: libs/dotnet/SquareSums/Metrics.cs
using System;
using System.Collections.Generic;

namespace SquareSums
{
    public class Metrics
    {
        private readonly bool _printProgress;
        private readonly Dictionary<int, int> _dfsCounterMap = new();
        
        private int _dfsCounter;
        
        public Metrics(bool printProgress)
        {
            _printProgress = printProgress;
        }

        public void IncrementDfsCounter()
        {
            _dfsCounter += 1;
        }

        public void FinalizeDfsCounter(int n)
        {
            _dfsCounterMap[n] = _dfsCounter;
            if (_printProgress)
            {
                Console.WriteLine($"Done: {n}: dfs counter {_dfsCounter}");
            }

            _dfsCounter = 0;
        }

        public void PrintMetrics()
        {
            var badCounter = 0;
            var normalCounter = 0;
            var averageCounter = 0;
            var maxN = 0;
            var worst = 0;
            var worstN = 0;
            foreach (var n in _dfsCounterMap.Keys)
            {
                if (n > maxN)
                {
                    maxN = n;
                }

                var dfsCounter = this._dfsCounterMap[n];
                if (dfsCounter > worst)
                {
                    worst = dfsCounter;
                    worstN = n;
                }

                if (dfsCounter / n > 3)
                {
                    //console.log(`Counter for ${n}: ${dfsCounter}`);
                    badCounter++;
                }
                else if (dfsCounter <= n)
                {
                    normalCounter++;
                }
                else
                {
                    averageCounter++;
                }
            }

            Console.WriteLine($"Normal (*<=n) dfs cases count: {normalCounter}/{maxN}");
            Console.WriteLine($"Average (n<*<3n) dfs cases count: {averageCounter}/{maxN}");
            Console.WriteLine($"Bad dfs (*>3n) cases count: {badCounter}/{maxN}");
            Console.WriteLine($"Worst dfs case: {worst} for {worstN}");
        }
    }
}

File: libs/rust/src/metrics.rs
use std::collections::HashMap;

pub struct Metrics {
    print_progress: bool,
    dfs_counter_map: HashMap<u32, u32>,
    dfs_counter: u32,
}

impl Metrics {
    pub fn new(print_progress: bool) -> Metrics {
        Metrics {
            print_progress,
            dfs_counter: 0,
            dfs_counter_map: HashMap::new(),
        }
    }

    pub fn increment_dfs_counter(&mut self) {
        self.dfs_counter += 1
    }

    pub fn finalize_dfs_counter(&mut self, n: u32) {
        self.dfs_counter_map.insert(n, self.dfs_counter);
        if self.print_progress {
            println!("Done: {}: dfs counter {}", n, self.dfs_counter);
        }

        self.dfs_counter = 0
    }

    pub fn print_metrics(&self) {
        let mut bad_counter = 0;
        let mut normal_counter = 0;
        let mut average_counter = 0;
        let mut max_n = 0;
        let mut worst = 0;
        let mut worst_n = 0;
        for (n, dfs_counter) in &self.dfs_counter_map {
            if *n > max_n {
                max_n = *n;
            }

            if *dfs_counter > worst {
                worst = *dfs_counter;
                worst_n = *n;
            }

            if dfs_counter / n > 3 {
                bad_counter += 1;
            } else if dfs_counter <= n {
                normal_counter += 1;
            } else {
                average_counter += 1;
            }
        }

        println!(
            "Normal (*<=n) dfs cases count: {}/{}",
            normal_counter, max_n
        );
        println!(
            "Average (n<*<3n) dfs cases count: {}/{}",
            average_counter, max_n
        );
        println!("Bad dfs (*>3n) cases count: {}/{}", bad_counter, max_n);
        println!("Worst dfs case: {} for {}", worst, worst_n);
    }
}
